cdc/owner/scheduler.go

```diff
@@ -0,0 +1,359 @@
+			TableID:    tableID,
+			BoundaryTs: boundaryTs,
+		})
+	}
+	// The remaining tables is the tables which should be not listened
+	tableShouldNotListened := allTableListeningNow
+	for tableID, captureID := range tableShouldNotListened {
+		opts := s.state.TaskStatuses[captureID].Operation
+		if opts != nil && opts[tableID] != nil && opts[tableID].Delete {
+			// the table is removing, skip
+			continue
+		}
+		pendingJob = append(pendingJob, &schedulerJob{
+			Tp:            schedulerJobTypeRemoveTable,
+			TableID:       tableID,
+			BoundaryTs:    globalCheckpointTs,
+			TargetCapture: captureID,
+		})
+	}
+	return pendingJob
+}
+
+func (s *scheduler) handleJobs(jobs []*schedulerJob) {
+	for _, job := range jobs {
+		job := job
+		s.state.PatchTaskStatus(job.TargetCapture, func(status *model.TaskStatus) (*model.TaskStatus, bool, error) {
+			switch job.Tp {
+			case schedulerJobTypeAddTable:
+				if status == nil {
+					// if task status is not found, we can just skip to set adding-table operation, this table will be added in next tick
+					log.Warn("task status of the capture is not found, may be the capture is already down. specify a new capture and redo the job", zap.Any("job", job))
+					return status, false, nil
+				}
+				status.AddTable(job.TableID, &model.TableReplicaInfo{
+					StartTs:     job.BoundaryTs,
+					MarkTableID: 0, // mark table ID will be set in processors
+				}, job.BoundaryTs)
+			case schedulerJobTypeRemoveTable:
+				failpoint.Inject("OwnerRemoveTableError", func() {
+					// just skip remove this table
+					failpoint.Return(status, false, nil)
+				})
+				if status == nil {
+					log.Warn("task status of the capture is not found, may be the capture is already down. specify a new capture and redo the job", zap.Any("job", job))
+					return status, false, nil
+				}
+				status.RemoveTable(job.TableID, job.BoundaryTs, false)
+			default:
+				log.Panic("Unreachable, please report a bug", zap.Any("job", job))
+			}
+			return status, true, nil
+		})
+	}
+}
+
+// cleanUpFinishedOperations clean up the finished operations.
+func (s *scheduler) cleanUpFinishedOperations() {
+	for captureID := range s.state.TaskStatuses {
+		s.state.PatchTaskStatus(captureID, func(status *model.TaskStatus) (*model.TaskStatus, bool, error) {
+			changed := false
+			for tableID, operation := range status.Operation {
+				if operation.Status == model.OperFinished {
+					delete(status.Operation, tableID)
+					changed = true
+				}
+			}
+			return status, changed, nil
+		})
+	}
+}
+
+func (s *scheduler) rebalance() (shouldUpdateState bool) {
+	if !s.shouldRebalance() {
+		return true
+	}
+	// we only support rebalance by table number for now
+	return s.rebalanceByTableNum()
+}
+
+func (s *scheduler) shouldRebalance() bool {
+	if s.needRebalanceNextTick {
+		s.needRebalanceNextTick = false
+		return true
+	}
+	if s.lastTickCaptureCount != len(s.captures) {
+		// a new capture online and no table distributed to the capture
```

Comment:
Is it possible for a capture to go out and another to come in, between the same two ticks?

Reply:
if a capture go out, the tables which listened in the offline capture will be dispatch to other captures in `syncTablesWithSchemaManager` function